modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -443,9 +444,15 @@ private CompletableFuture<AsyncSqlCursor<List<Object>>> querySingle0(
 
                     boolean implicitTxRequired = outerTx == null;
 
-                    tx.set(implicitTxRequired ? txManager.begin(!rwOp, null) : outerTx);
+                    InternalTransaction currentTx = implicitTxRequired ? txManager.begin(!rwOp, null) : outerTx;
 
-                    SchemaPlus schema = sqlSchemaManager.schema(schemaName);
+                    tx.set(currentTx);
+
+                    // TODO IGNITE-18733: wait for actual metadata for TX.
+                    HybridTimestamp txTimestamp = currentTx.startTimestamp();
+                    int observableCatalogVersion = catalogManager.activeCatalogVersion(txTimestamp.longValue());
+
+                    SchemaPlus schema = sqlSchemaManager.schema(schemaName, observableCatalogVersion);
```

Comment:
This is where I get confused too.
We get the catalog version and use it as the schema version, which may not be true.
Maybe we need, by analogy with the catalog, to wait for the schema version by `txTimestamp`?

Reply:
Ok, we wait for catalog version and want to get corresponding SQL schema.
SchemaPlus is Calcite class and we can' rename it.
I can add a comment.... WDYT?

Reply:
Whether correctly I understand that the version of the scheme and the catalog are equal?

If yes, then it is necessary to correct the documentation of the method and the naming of variables, for example, obtaining the schema by the name and version of the directory.

If not, then you probably need to fix or create a ticket (indicate the existing one) that this will be fixed in the future.

At the moment, I assume that the versions of the schemas and the catalog will be different, since an entity not related to the schema, such as a zone, table, and index, may change in the catalog.

Reply:
1. `SqlSchemaManager` is a general interface, which provides method to get schema by name and version. It doesn't assumed any connection with Catalog, and whether version should match catalog version or not. 
I think we shouldn't fix any naming here.
2. `CatalogSqlSchemaManager is an implementation that backed by catalog. We assume schema version is mapped to catalog version as 1:1.
3. Actually, current implementation `SqlSchemaManager` does not support versions at all. So, any naming is possible here, as it make no sense.

Maybe, we should either use `CatalogSqlSchemaManager` as `sqlSchemaManager` type, or add a method to `SqlSchemaManager` interface that will resolve a version by tx timestamp. WDYT?

Reply:
> Maybe, we should either use CatalogSqlSchemaManager as sqlSchemaManager type, or add a method to SqlSchemaManager interface that will resolve a version by tx timestamp. WDYT?

Yes, I like the option with an additional method.